File: leetcode/go/src/leetcode/114_flatten_binary_tree_to_linked_list.go
package leetcode

// flatten a inner TreeNode and return tail
func innerFlatten(root *TreeNode) *TreeNode {
	if root == nil || (root.Left == nil && root.Right == nil) {
		return root
	}
	l, r := root.Left, root.Right
	var ret *TreeNode
	if l != nil {
		lt := innerFlatten(l)
		ret = lt
		root.Right = l
		root.Left = nil
		lt.Right = r
	}
	if r != nil {
		rt := innerFlatten(r)
		ret = rt
	}
	return ret
}

func flatten(root *TreeNode) {
	innerFlatten(root)
}

File: leetcode/cpp/114_flatten_binary_tree_to_linked_list.cpp
class Solution {
 public:
  // time: O(n), space: O(1)
  void flatten(TreeNode* root) {
    if (root == nullptr) return;
    auto left = root->left;
    auto right = root->right;
    flatten(left);
    flatten(right);
    root->left = nullptr;
    root->right = left;

    auto temp = left;
    while (temp != nullptr && temp->right != nullptr) {
      temp = temp->right;
    }
    if (temp != nullptr) {
      temp->right = right;
    } else {
      root->right = right;
    }
  }
};
